src/util/sharing_map.h

```diff
@@ -1233,88 +1299,93 @@ ::insert(const key_type &k, valueU &&m)
     SM_ASSERT(ip->is_internal());
     SM_ASSERT(level == 0 || !ip->empty());
 
-    innert *child = ip->add_child(bit);
+    nodet &child = ip->add_child(bit);
 
     // Place is unoccupied
-    if(child->empty())
+    if(child.empty())
     {
-      // Create container and insert leaf
-      child->place_leaf(k, std::forward<valueU>(m));
+      if(level < levels - 1)
+      {
+        // Create leaf
+        child.make_leaf(k, m);
+      }
+      else
+      {
+        SM_ASSERT(level == levels - 1);
 
-      SM_ASSERT(child->is_defined_container());
+        // Create container and insert leaf
+        child.place_leaf(k, std::forward<valueU>(m));
+
+        SM_ASSERT(child.is_defined_container());
+      }
 
       num++;
 
       return;
     }
-
-    if(child->is_container())
+    else if(child.is_internal())
     {
-      if(level < levels - 1)
-      {
-        // Migrate the elements downwards
-        innert *cp = migrate(level, key, bit, *ip);
+      ip = &child;
+      key >>= chunk;
+      level++;
 
-        cp->place_leaf(k, std::forward<valueU>(m));
-      }
-      else
-      {
-        // Add to the bottom container
-        child->place_leaf(k, std::forward<valueU>(m));
-      }
+      continue;
+    }
+    else if(child.is_leaf())
+    {
+      // migrate leaf downwards
+      migrate(level, key, bit, *ip, k, std::forward<valueU>(m));
 
       num++;
 
       return;
     }
+    else
+    {
+      SM_ASSERT(child.is_defined_container());
+      SM_ASSERT(level == levels - 1);
 
-    SM_ASSERT(level == levels - 1 || child->is_defined_internal());
+      // Add to the container
+      child.place_leaf(k, std::forward<valueU>(m));
 
-    ip = child;
-    key >>= chunk;
-    level++;
+      num++;
+
+      return;
+    }
   }
 }
 
 SHARING_MAPT4(valueU, void)
 ::replace(const key_type &k, valueU &&m)
 {
-  innert *cp = get_container_node(k);
-  SM_ASSERT(cp != nullptr);
-
-  leaft *lp = cp->find_leaf(k);
-  PRECONDITION(lp != nullptr); // key must exist in map
+  nodet &lp = get_leaf_node(k);
 
   INVARIANT(
-    !value_equalt()(as_const(lp)->get_value(), m),
+    !value_equalt()(as_const(lp).get_value(), m),
```

Comment:
Visual Studio complains that the use of `as_const` is ambiguous here. You might have add/use a static cast here.